File: src/model/rules/RuleImplGomoku.java
package model.rules;

import model.rules.displayRule.DisplayRule;
import model.rules.displayRule.DisplayRuleImplGomoku;
import model.rules.gameRule.GameRule;
import model.rules.gameRule.GameRuleImplGomoku;
import model.rules.inputRule.InputRule;
import model.rules.inputRule.InputRuleImplMonochrome;

public class RuleImplGomoku implements Rule {

    private static final RuleImplGomoku instance = new RuleImplGomoku();

    public static RuleImplGomoku getRule() {
        return instance;
    }

    private RuleImplGomoku() {}

    final InputRule inputRule = InputRuleImplMonochrome.getInputRule();
    final GameRule gameRule = GameRuleImplGomoku.getGameRule();
    final DisplayRule displayRule = DisplayRuleImplGomoku.getDisplayRule();

    @Override
    public GameRule getGameRule() {
        return gameRule;
    }

    @Override
    public InputRule getInputRule() {
        return inputRule;
    }

    @Override
    public DisplayRule getDisplayRule() {
        return displayRule;
    }

    public static final String name ="gomoku";
    @Override
    public String getName() { return name; }
}

File: src/model/rules/RuleImplGomoku.kt
package model.rules

import model.rules.displayRule.DisplayRule
import model.rules.displayRule.DisplayRuleImplGomoku
import model.rules.gameRule.GameRule
import model.rules.gameRule.GameRuleImplGomoku
import model.rules.inputRule.InputRule
import model.rules.inputRule.InputRuleImplMonochrome

class RuleImplGomoku private constructor() : Rule {
    override val inputRule: InputRule = InputRuleImplMonochrome.inputRule
    override val gameRule: GameRule = GameRuleImplGomoku.gameRule
    override val displayRule: DisplayRule = DisplayRuleImplGomoku.displayRule

    override val name: String
        get() = Companion.name

    companion object {
        val rule: RuleImplGomoku = RuleImplGomoku()

        const val name: String = "gomoku"
    }
}
